spotbugs/src/main/java/edu/umd/cs/findbugs/detect/ReflectionIncreaseAccessibility.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * SpotBugs - Find bugs in Java programs
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+package edu.umd.cs.findbugs.detect;
+
+import org.apache.bcel.Const;
+import org.apache.bcel.classfile.JavaClass;
+import org.apache.bcel.classfile.Method;
+
+
+import edu.umd.cs.findbugs.BugAccumulator;
+import edu.umd.cs.findbugs.BugInstance;
+import edu.umd.cs.findbugs.BugReporter;
+import edu.umd.cs.findbugs.OpcodeStack;
+import edu.umd.cs.findbugs.ba.AnalysisContext;
+import edu.umd.cs.findbugs.ba.XMethod;
+import edu.umd.cs.findbugs.bcel.OpcodeStackDetector;
+
+public class ReflectionIncreaseAccessibility extends OpcodeStackDetector {
+    private boolean securityCheck = false;
+    private boolean fieldNameUnderGet = false;
+
+    private final BugAccumulator bugAccumulator;
+
+    public ReflectionIncreaseAccessibility(BugReporter bugReporter) {
+        this.bugAccumulator = new BugAccumulator(bugReporter);
+    }
+
+    @Override
+    public void visit(Method met) {
+        securityCheck = false;
+        super.visit(met);
+    }
+
+    @Override
+    public void visitAfter(JavaClass obj) {
+        bugAccumulator.reportAccumulatedBugs();
+    }
+
+    @Override
+    public void sawOpcode(int seen) {
+        fieldNameUnderGet = false;
+
+        if (seen != Const.INVOKEVIRTUAL || !getThisClass().isPublic() || !getMethod().isPublic()) {
+            return;
+        }
+
+        OpcodeStack.Item obj = stack.getItemMethodInvokedOn(this);
+        try {
+            JavaClass cls = obj.getJavaClass();
+            if (cls == null) {
+                return;
+            }
+            XMethod met = getXMethodOperand();
+            if (!securityCheck && obj.isInitialParameter() && obj.getXField() == null &&
+                    "java.lang.Class".equals(cls.getClassName()) &&
+                    "newInstance".equals(met.getName()) &&
+                    "()Ljava/lang/Object;".equals(met.getSignature())) {
```

Comment:
As [I commented previously](https://github.com/spotbugs/spotbugs/pull/1541#discussion_r735461270https://github.com/spotbugs/spotbugs/pull/1541#discussion_r735461270), it is necessary to check visibility of target class and current method. For instance, if both are public then there is no vulnerability.

Reply:
Sorry, I was probably not clear enough when I [tried to explain the purpose of the detector](https://github.com/spotbugs/spotbugs/pull/1541#issuecomment-950813189). We cannot check the visibility of the target class here because we do not have the attacker code. The issue is that our function (which is public and is defined in a public class) takes the class as a parameter thus any class can be passed to it and our function instantiates them. This is a security vulnerability. Thus we do not know the value of the parameter (and would not know it even if the attacker code is present during the analysis [but it is of course not] because SpotBugs is not interprocedural in this way) at the time of the analysis we only know that there is risk that someone passes a class to the function which may not have a public constructor thus it is not intended to be instantiated by anyone. The same applies to fields as well which may be private but our function sets their value or sets the accessibility to true.